File: slidingWindows/maxConsecutive.js
/*
Given a binary array nums and an integer k, return the maximum number of consecutive 1's in the array if you can flip at most k 0's.

Example 1:

Input: nums = [1,1,1,0,0,0,1,1,1,1,0], k = 2
Output: 6
Explanation: [1,1,1,0,0,1,1,1,1,1,1]
Bolded numbers were flipped from 0 to 1. The longest subarray is underlined.

Example 2:

Input: nums = [0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1], k = 3
Output: 10
Explanation: [0,0,1,1,1,1,1,1,1,1,1,1,0,0,0,1,1,1,1]
Bolded numbers were flipped from 0 to 1. The longest subarray is underlined.

Constraints:

    1 <= nums.length <= 105
    nums[i] is either 0 or 1.
    0 <= k <= nums.length

*/

/**
 * 
 * @param {Array} arr array to work with
 * @param {int} k size of the fixed window
 */
function getMaxConsecutiveOne(arr,k){
    let maxSize = 0;//The longest consecutive subarray of 1 (with k 0 flipped)
    let l = 0; //initial position of l (the left side of the sliding window)
    let usedZeros = 0; //number of zeroes we used
    let w = 0; //weight/size of the current window

    for(let r = 0;r<arr.length;r++){
        //If the r side finds a 0, convert it 
        if(arr[r] === 0){
            usedZeros++; 
        }

        //if we have more zeros than allowed(k) move l until this get solved!
        while(usedZeros > k){
            if(arr[l] == 0){
                usedZeros--;
            }
            l++;        
        }

        w = (r-l)+1;

        if(w > maxSize){
            maxSize = (r-l)+1;
        }
    }
    
    return maxSize;

}


let testInput = [1,1,1,0,0,0,1,1,1,1,0];
console.log(getMaxConsecutiveOne(testInput,2));

File: slidingWindows/maxConsecutive.cpp
#include <iostream>
#include <vector>
#include <string>

int maxConsecutive(std::vector<int>& bin, int& k){
    int l = 0;//left position of window
    int w = 0;//weight/size of window
    int usedFlips = 0;//number of used flips
    int maxConsecutive = 0;//max string of 1's 
    
    for(int r=0;r<bin.size();r++){

        if(bin[r] == 0){
            usedFlips = usedFlips + 1;
        }

        while(usedFlips > k){
            if(bin[l] == 0){
                usedFlips = usedFlips - 1;
            }
            l = l + 1;
        }

        w = (r-l)+1;

        if(w>maxConsecutive){
            maxConsecutive = w;
        }

    }

    return maxConsecutive;
}

